Add foundation & initial implementation for a full-blown wallet CLI

jonathanknowles opened the pull request:
Experimental!

# Issue Number

#96 

# Overview

Add beginnings of a docopt-based CLI for the Wallet.

KtorZ commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -1370,81 +1368,6 @@ instance FromJSON AddressLevel where
-            Just (mkLevel n)
-    AccountIndex actualIndex = accIndex acct
-
-
```
:+1:

KtorZ commented on src/Cardano/Wallet/Kernel/Ed25519Bip44.hs:
```diff
@@ -0,0 +1,35 @@
+deriveHDPublicKeyScheme2 :: PublicKey -> Word32 -> Maybe PublicKey
+deriveHDPublicKeyScheme2 (PublicKey xpub) childIndex =
+    maybe Nothing
+          (Just . PublicKey)
```
`fmap` ?

KtorZ commented on src/Cardano/Wallet/Kernel/Ed25519Bip44.hs:
```diff
@@ -0,0 +1,35 @@
+    -> Word
+    -> PublicKey
+deriveLvl3PublicKey accountPublicKey addressIx =
+    maybe (error "Unable to derive address public key from account public key!")
```
Let's not throw in the wild like this. Especially in this module. Leave that to the caller.

KtorZ commented on src/Cardano/Wallet/Kernel/Ed25519Bip44.hs:
```diff
@@ -0,0 +1,35 @@
+-- | Derives extended public key from extended public key using derivation scheme 2
+-- (please see 'cardano-crypto' package).
+deriveHDPublicKeyScheme2 :: PublicKey -> Word32 -> Maybe PublicKey
+deriveHDPublicKeyScheme2 (PublicKey xpub) childIndex =
```
No need for this indirection, let just inline this function in the call above.

akegalj commented on src/Cardano/Wallet/Kernel/Ed25519Bip44.hs:
```diff
@@ -0,0 +1,26 @@
+
+-- | Derives address public key from the given account public key,
+-- using derivation scheme 2 (please see 'cardano-crypto' package).
+deriveLvl3PublicKey
```
@denisshevchenko @KtorZ 
Hm, I don't think this is right. If we are about to implement bip44 - derivation path is defined as
```
m / purpose' / coin_type' / account' / change / address_index
```
if we are providing account public key here, we still have to first derive internal/external change derivation before deriving an address.

Or we decided not to bother with change derivation in our implementation?

I am not sure do we have a notion of internal/external wallet atm (v0 had this adhoc) 

Also, if we are implementing bip44 , why not sticking with proposed naming convention:

m => level 0
purpose' => level 1
coin_type' => level 2
account' => level 3
change => level 4
address_index => level 5 (instead of level 3)

KtorZ replied:
We don't manipulate derivation path so to speak. `cardano-crypto` does it. We only use the abstract type known as `PublicKey` / `PrivateKey`. The rest is irrelevant to us as we aren't implementing the cryptographic details of it.

akegalj replied:
its a shame wallet can't fetch change derivation index as this would be useful information (to show an address or to hide it; it would also be useful for explorer).

NicolasDP replied:
@KtorZ `cardano-crypto` does not implement the Bip44 scheme. only the ed25519-bip32 derivation scheme (which is a cryptographic primitive). You have to implement the bip44 scheme on your side of the wallet (i.e. hardened derivation for purpose and coin_type and account, and then soft derivation for change and address).

so like @akegalj said, if you have the account public key, you need to derive your internal (or external) first and then derive the address key. 2 derivation levels.

There is no understanding of path in the cardano-crypto library, only the derivation index from one public key to another. so you need to implement every derivation levels of the BIP44 on your side.

KtorZ replied:
Indeed. Bad choice of wording on my side. What I meant was that, there's no need to define "paths" on our side and keep track of those, so to speak. Paths are just another round of derivation. We just need to provide utils that perform the correct number of derivations for the right type given the corresponding address or account indexes (for intermediate derivation like purpose and coin_type, values are just hardcoded). In the end, we only manipulate `XPub` and `XPrv` keys.

From an external user, this looks roughly like:

```hs
deriveAccountPrv :: (XPrv, PassPhrase) -> Word32 -> XPrv
   -- perform triple derivation from the given wallet XPrv 
deriveAddressPub :: XPub -> IsChange -> Word32 -> XPub
   -- perform double derivation from the given account XPub 
```

akegalj replied:
yes - seems like there was a noise in communication. Thanks for having a look @NicolasDP

akegalj commented on test/unit/Test/Spec/Ed25519Bip44.hs:
```diff
@@ -0,0 +1,36 @@
+    result = deriveLvl3PublicKey accountPublicKey addressIx
+    -- This is maximum value for soft derivation (only soft derivation
+    -- is allowed to derive public key from public key).
+    maxIx = 0x8000000
```
@denisshevchenko I guess you missed one zero here. It should be 0x80000000 (seven zeros).

(the test is failing then :D)

akegalj replied:
but it seems to be a bug in cardano-crypto, not in your code (or I am not understanding bip32 well) - will have a closer look at it

akegalj replied:
When single zero is dropped as @denisshevchenko did it, it is actually the same boundery where `deriveXPub` is failing:
```
> logBase 2 0x8000000
27.0
```

so either both @denisshevchenko and `deriveXPub` are correct and I am missing some crutial understanding of bip32 https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#extended-keys or when 0 zero is added we get:
```
> logBase 2 0x8000000
31.000000000000004
```

which is actually boundary defined for hardened keys https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#extended-keys

It seems very unlikely that both Denis and crypto is wrong. Looks more likely that I am missing some info?

akegalj replied:
bip32 https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#public-parent-key--public-child-key also doesn't reveal why `deriveXPub` should work only for `<2^27` . Having a look at `cardano-crypto` internals a bit...

akegalj replied:
Fix should be here https://github.com/input-output-hk/cardano-crypto/pull/52

jonathanknowles committed:
Add the beginnings of a docopt-based CLI for the Wallet.

KtorZ commented on app/wallet/USAGE.txt:
```diff
@@ -0,0 +1,14 @@
+
+Usage:
+  cardano-wallet address list --wallet-id=UUID
+  cardano-wallet wallet create --name=STRING --mnemonic-sentence=STRING --passphrase=STRING [--address-pool-gap=INT] [--mnemonic-second-factor=STRING]
```
The mnemonics will have to be prompted :no_mouth: These are sensitive pieces of info we do not want in the bash history :sweat_smile:

jonathanknowles replied:
> The mnemonics will have to be prompted no_mouth These are sensitive pieces of info we do not want in the bash history sweat_smile

That's a really good point. I'm happy to add this. Is there any particular approach you would prefer for reading the mnemonics in?

KtorZ replied:
Just a plain shell prompt, doesn't necessarily have to be hidden (because it may become very tricky to type-in 24 words if you can see shit :joy: ) 
But clearing the prompt once done.

```
$ Please enter your 12-to-24-word mnemonics: ...
```
|
|
v

```
$ Please enter your 12-to-24-word mnemonics: patate patate patate patate
```
|
| 
v

hits enter

|
|
v
```
Please enter your 12-to-24 word mnemonics: ****
Thank you, now we've got all your ADA in a safe place
```

jonathanknowles replied:
> The mnemonics will have to be prompted :no_mouth: 
> These are sensitive pieces of info we do not want in the bash history :sweat_smile: 
> But clearing the prompt once done.

@KtorZ  I've now implemented a basic version of the above. It's still a bit rough at the edges (in particular, with the reporting of errors), but clearing the prompt does work.

One issue that I ran into when clearing data from a terminal screen is that it's not always possible to determine the original position of the cursor. If we're not careful, we run the risk of not clearing all the sensitive data from the display.

Therefore, if the system on which the CLI is running is not able to provide this info, I've opted to clear the entire screen after the user has finished entering their data.

KtorZ commented on app/wallet/Main.hs:
```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Main where
```
Side-note but this cli and `cardano-wallet-server` are one and one thing. We do want to have them merged together in the end, so that there's one CLI to spawn the server, and the same one to interact with it :+1:

jonathanknowles replied:
> Side-note but this cli and `cardano-wallet-server` are one and one thing. We do want to have them merged together in the end, so that there's one CLI to spawn the server, and the same one to interact with it +1

Understood! In that case, are you expecting something like this:

`cardano-wallet server start ...`
`cardano-wallet wallet create --name wibble ...`
`cardano-wallet wallet list`

KtorZ replied:
Sounds reasonable to me :)

KtorZ commented on src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -202,91 +199,111 @@ getApiMnemonicT (ApiMnemonicT (pw, _)) = pw
+
+-- | Indicates an error that occurred while decoding from text.
+newtype TextDecodingError = TextDecodingError
+    { getTextDecodingError :: String }
```
Not sure what's the value of the `newtype` here. I'd simply go for a string to be honest. It adds value when we have multiple ones, one per actual type (albeit having a raw string isn't really the best option). But since it has now become a common abstraction, a `String` has pretty much the same value as the newtype here.

jonathanknowles replied:
I think there is some value in having a `newtype` here.

1. The descriptive name of the type `TextDecodingError` is self-commenting. When I see this type in the wild somewhere, I don't need to think very hard to know that it's an error relating to the decoding of text, and not some random string that happens to have been inserted into the left-hand-side of an `Either`.

2. If I see `TextDecodingError` in the wild, in a place _outside_ of the `Api.Types` module, then I have a direct way to look up its definition, at which point I can see the extended comment.

3. If I were to try refactor `TextDecodingError` in some way, then I might get better compiler feedback than if it were just a naked `String`.

What do you think? Of course, if you think the arguments I've presented have no basis, then we can remove it.

(Note that I've now created a separate PR for the textual encoding work: #142.)

KtorZ commented on src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -38,15 +38,9 @@ module Cardano.Wallet.Api.Types
-    , encodeApiWalletName
+    , ToText (..)
+    , FromText (..)
+    , TextDecodingError (..)
```
The overall looks clean and re-usable. Nice work! 
(one reserve about the `TextDecodingError`)

jonathanknowles replied:
Thanks! I've moved the textual encoding work into a separate PR: #142.

I'll keep this PR for the `docopt` CLI-specific work.

jonathanknowles committed:
Add function `getLineWithSensitiveData`.

This function reads a line containing sensitive data from the terminal.

The terminal lines containing the data are cleared once the user has
finished entering their data.

jonathanknowles committed:
Add structure to the docopt-based CLI module.

jonathanknowles committed:
Add functionality to gather sensitive data for wallet creation.

KtorZ committed:
unify CLIs approaches and CLI themselves into on 'cardano-wallet' CLI

KtorZ committed:
move 'FromText' and 'ToText' into their own modules

KtorZ committed:
move 'FromText' & 'ToText' instance closer to their corresponding types

KtorZ reviewed (approved):
Reviewed & Wrapped up.
I'll add a few more tests and merge it.

akegalj commented on app/Cardano/CLI.hs:
```diff
@@ -12,103 +11,145 @@
+-- The terminal lines containing the data are also cleared if the application
+-- exits abnormally, before the user has finished entering data.
+--
+getLineWithSensitiveData :: IO Text
```
alternitavelly we could even hide user input from the screen (similar like `sudo` does). 

The cons of hiding the user input is that user is unable to check/verify/easily-correct entered mnemonic or password.

Some CLI application approach the problem of requesting the same input data to be done twice in order to confirm that user didn't make a mistake:

```
Passphrase: ********
Retype passphrase: ********
```
user experience is lower with this model but users are now able to enter their their mnemonics or passphrases in public spaces.

Mnemonics are pretty long (15+ words) and passphrase as well (10+ chars) so there is a high chance of mistake. That's also the case not to implement what I am proposing.

I guess CLI will most likely be used by someone already more technically advanced - so there is a higher chance that person already is aware that mnemonic and passphrase is something he should hide and not reveal (ie, type in public place).

Conclusion is that I think this is ok solution but maybe we should keep in mind and research alternatives.
